parking_lot.py

```diff
@@ -0,0 +1,172 @@
+
+
+@app.exception_handler(LocationNotAvailableException)
+async def location_not_available_exception_handler(exception: LocationNotAvailableException):
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Cannot park your car there because location {exception.location} is not available!"})
+
+
+@app.exception_handler(StartDateException)
+async def start_date_in_the_future(exception: StartDateException):
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Cannot part your car because date {exception.start} is invalid!"})
+
+
+@app.exception_handler(CarIdExistsException)
+async def car_id_already_exists(exception: CarIdExistsException):
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Car {exception.car_id} is already parked!"})
+
+
+@app.exception_handler(CarIdDoesNotExistException)
+async def car_id_does_not_exist(exception: CarIdDoesNotExistException):
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Car {exception.car_id} is not in the parking lot!!"})
+
+
+@app.exception_handler(CarAlreadyRemovedException)
+async def car_id_already_removed(exception: CarAlreadyRemovedException):
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Car {exception.car_id} already removed!"})
+
+
+@app.get("/cars/")
+async def get_cars():
+    return PARKED_CARS
+
+
+@app.get("/cars/{car_id}", status_code=status.HTTP_200_OK)
+async def get_car(car_id: Optional[str] = None):
+    return PARKED_CARS[car_id]
+
+
+@app.post("/cars/", status_code=status.HTTP_201_CREATED)
+async def add_car(parked_car: ParkedCar):
+    if parked_car.location <= 0 or parked_car.location > PARKING_LOT_CAPACITY:
+        raise InvalidLocationException(location=parked_car.location)
+
+    if parked_car.location in [x.get('location') for x in PARKED_CARS.values()]:
+        raise LocationNotAvailableException(location=parked_car.location)
+
+    start_date = parked_car.start.replace(tzinfo=None)
+    present = datetime.now()
+
+    if (present - start_date).total_seconds() > ACCEPTED_TIME_DELAY:
+        raise StartDateException(start=parked_car.start)
+
+    if parked_car.car_id in PARKED_CARS.keys():
+        raise CarIdExistsException(car_id=parked_car.car_id)
+
+    PARKED_CARS[parked_car.car_id] = {'status': 'parked', 'tariff': parked_car.tariff, 'location': parked_car.location, 'start': parked_car.start, 'end': None, 'fee': None}
+    return PARKED_CARS[parked_car.car_id]
+
+
+@app.put("/cars/{car_id}", status_code=status.HTTP_200_OK)
+async def remove_car(car_id: str, parked_car: ParkedCar):
+    if car_id not in PARKED_CARS.keys():
+        raise CarIdDoesNotExistException(car_id=car_id)
+
+    if PARKED_CARS[car_id].get('status') == 'removed':
+        raise CarAlreadyRemovedException(car_id=car_id)
```

Comment:
I'll change the location to 'location': '' on car remove

Reply:
From my point of view, each parking record should be preserved, in this case, if the car is reparked, older records are lost. The tracking of parking could be based on record_id not just car_id

Reply:
It can either from db or memory, we don't have to use db, we can still use memory but the logic is the same. Memory vs db is just about expensive/cheap storage, but can with same logics.